File: rayt-rust/src/vec3.rs
use crate::{random_double, utils::PI};
use std::fmt;
use std::ops::{Add, Div, Mul, Neg, Sub};

#[derive(PartialEq, Clone, Copy, Default)]
pub struct Vec3 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

impl Vec3 {
    pub fn random(min_max: Option<[f64; 2]>) -> Self {
        Self {
            x: random_double!(min_max),
            y: random_double!(min_max),
            z: random_double!(min_max),
        }
    }

    pub fn length(self) -> f64 {
        f64::sqrt(self.length_squared())
    }

    pub fn length_squared(self) -> f64 {
        self.x.powi(2) + self.y.powi(2) + self.z.powi(2)
    }
}

impl fmt::Debug for Vec3 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Vec3 ({:?}, {:?}, {:?})", self.x, self.y, self.z)
    }
}

impl From<[f64; 3]> for Vec3 {
    fn from(e: [f64; 3]) -> Self {
        Self {
            x: e[0],
            y: e[1],
            z: e[2],
        }
    }
}

impl Add<Vec3> for Vec3 {
    type Output = Vec3;

    fn add(self, rhs: Vec3) -> Self::Output {
        Self::Output {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }
}

impl Sub<Vec3> for Vec3 {
    type Output = Vec3;

    fn sub(self, rhs: Vec3) -> Self::Output {
        Self::Output {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl Mul<Vec3> for Vec3 {
    type Output = Vec3;

    fn mul(self, rhs: Vec3) -> Self::Output {
        Self::Output {
            x: self.x * rhs.x,
            y: self.y * rhs.y,
            z: self.z * rhs.z,
        }
    }
}

impl Mul<f64> for Vec3 {
    type Output = Vec3;

    fn mul(self, rhs: f64) -> Self::Output {
        Self::Output {
            x: self.x * rhs,
            y: self.y * rhs,
            z: self.z * rhs,
        }
    }
}

impl Mul<Vec3> for f64 {
    type Output = Vec3;

    fn mul(self, rhs: Vec3) -> Self::Output {
        rhs * self
    }
}

impl Div<Vec3> for Vec3 {
    type Output = Vec3;

    fn div(self, rhs: Vec3) -> Self::Output {
        Self::Output {
            x: self.x / rhs.x,
            y: self.y / rhs.y,
            z: self.z / rhs.z,
        }
    }
}

impl Div<f64> for Vec3 {
    type Output = Vec3;

    fn div(self, rhs: f64) -> Self::Output {
        self * (1.0 / rhs)
    }
}

impl Neg for Vec3 {
    type Output = Vec3;

    fn neg(self) -> Self::Output {
        Self::Output {
            x: -self.x,
            y: -self.y,
            z: -self.z,
        }
    }
}

// Type aliases for Vec3
pub type Point3 = Vec3; // 3D point
pub type Color = Vec3; // RGB color

// Vec3 utility functions
pub fn dot(u: &Vec3, v: &Vec3) -> f64 {
    u.x * v.x + u.y * v.y + u.z + v.z
}

pub fn cross(u: Vec3, v: Vec3) -> Vec3 {
    Vec3::from([
        u.y * v.z - u.z * v.y,
        u.z * v.x - u.x * v.z,
        u.x * v.y - u.y * v.x,
    ])
}

pub fn unit_vector(v: Vec3) -> Vec3 {
    v / v.length()
}

pub fn random_in_unit_sphere() -> Vec3 {
    loop {
        let p = Vec3::random(Some([-1.0, 1.0]));
        if p.length_squared() >= 1.0 {
            continue;
        }
        return p;
    }
}

pub fn random_unit_vector() -> Vec3 {
    let a = random_double!(0.0, 2.0 * PI);
    let z = random_double!(-1.0, 1.0);
    let r = f64::sqrt(1.0 - z.powi(2));
    Vec3::from([r * f64::cos(a), r * f64::sin(a), z])
}

pub fn random_in_unit_disk() -> Vec3 {
    loop {
        let p = Vec3::from([random_double!(-1.0, 1.0), random_double!(-1.0, 1.0), 0.0]);
        if p.length_squared() >= 1.0 {
            continue;
        }
        return p;
    }
}

pub fn reflect(v: Vec3, n: Vec3) -> Vec3 {
    v - 2.0 * dot(&v, &n) * n
}

pub fn refract(uv: Vec3, n: Vec3, etai_over_etat: f64) -> Vec3 {
    let cos_theta = dot(&-uv, &n);
    let r_out_parallel = etai_over_etat * (uv + cos_theta * n);
    let r_out_perp = -f64::sqrt(1.0 - r_out_parallel.length_squared()) * n;
    r_out_parallel + r_out_perp
}

File: src/rayt/vec3.py
import math
import random
from typing import List, Optional, Union


class Vec3:
    def __init__(self, x: float = 0.0, y: float = 0.0, z: float = 0.0):
        self.x = x
        self.y = y
        self.z = z

    @classmethod
    def random(
        cls, min_max: Optional[Union[List[float], tuple[float, float]]] = None
    ) -> "Vec3":
        if min_max is None:
            return cls(random.random(), random.random(), random.random())
        else:
            min_val, max_val = min_max
            return cls(
                random.uniform(min_val, max_val),
                random.uniform(min_val, max_val),
                random.uniform(min_val, max_val),
            )

    @property
    def length(self) -> float:
        return math.sqrt(self.length_squared)

    @property
    def length_squared(self) -> float:
        return self.x**2 + self.y**2 + self.z**2

    def __add__(self, other: "Vec3") -> "Vec3":
        return Vec3(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other: "Vec3") -> "Vec3":
        return Vec3(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, other: Union["Vec3", float]) -> "Vec3":
        if isinstance(other, Vec3):
            return Vec3(self.x * other.x, self.y * other.y, self.z * other.z)
        else:
            return Vec3(self.x * other, self.y * other, self.z * other)

    def __rmul__(self, other: float) -> "Vec3":
        return self * other

    def __truediv__(self, other: Union["Vec3", float]) -> "Vec3":
        if isinstance(other, Vec3):
            return Vec3(self.x / other.x, self.y / other.y, self.z / other.z)
        else:
            return self * (1.0 / other)

    def __neg__(self) -> "Vec3":
        return Vec3(-self.x, -self.y, -self.z)

    def __iadd__(self, other: "Vec3") -> "Vec3":
        self.x += other.x
        self.y += other.y
        self.z += other.z
        return self

    def __imul__(self, other: Union["Vec3", float]) -> "Vec3":
        if isinstance(other, Vec3):
            self.x *= other.x
            self.y *= other.y
            self.z *= other.z
        else:
            self.x *= other
            self.y *= other
            self.z *= other
        return self

    def __repr__(self) -> str:
        return f"Vec3({self.x}, {self.y}, {self.z})"


# Type aliases
Point3 = Vec3
Color = Vec3


# Utility functions
def dot(u: Vec3, v: Vec3) -> float:
    return u.x * v.x + u.y * v.y + u.z * v.z


def cross(u: Vec3, v: Vec3) -> Vec3:
    return Vec3(u.y * v.z - u.z * v.y, u.z * v.x - u.x * v.z, u.x * v.y - u.y * v.x)


def unit_vector(v: Vec3) -> Vec3:
    return v / v.length


def random_in_unit_sphere() -> Vec3:
    while True:
        p = Vec3.random([-1.0, 1.0])
        if p.length_squared >= 1.0:
            continue
        return p


def random_unit_vector() -> Vec3:
    a = random.uniform(0.0, 2.0 * math.pi)
    z = random.uniform(-1.0, 1.0)
    r = math.sqrt(1.0 - z**2)
    return Vec3(r * math.cos(a), r * math.sin(a), z)


def reflect(v: Vec3, n: Vec3) -> Vec3:
    return v - 2.0 * dot(v, n) * n


def refract(uv: Vec3, n: Vec3, etai_over_etat: float) -> Vec3:
    cos_theta = dot(-uv, n)
    r_out_parallel = etai_over_etat * (uv + cos_theta * n)
    r_out_perp = -math.sqrt(1.0 - r_out_parallel.length_squared) * n
    return r_out_parallel + r_out_perp


def random_double(min_val: float = 0.0, max_val: float = 1.0) -> float:
    return random.uniform(min_val, max_val)
